Bug fix to samtools fastq

jenniferliddle opened the pull request:
Allow santools fastq to have the -i and --index-format options without
having to specify index files.

Fix bug whereby only the second index of a dual index tag was
being put into the -i (Casava) field.

This fixes the problems mentioned in #1047

daviesrob commented:
I'm not sure if this works properly yet.  `samtools fastq -i --index-format 'i4i4' ...` doesn't write out the extra tags, but `samtools fastq -i -i1 index1.fq --index-format 'i4i4' ...` does.  Also `samtools fastq -i ...` without `--index-format` does not complain about the lack of that option while it will if `-i1` is used.

When I do get them, the tags look like `1:N:0:ATCA-CGAT` on both reads, but [the bcl2fastq user manual](https://support.illumina.com/content/dam/illumina-support/documents/documentation/software_documentation/bcl2fastq/bcl2fastq2-v2-20-software-guide-15051736-03.pdf) says it should start with `2:` on read 2.  Unfortunately it doesn't say exactly what it does with dual-index runs, so I can't tell if the barcode bit is correct.  Do we have the output from bcl2fastq on one of these runs to we can see what it actually does?

Also (and somewhat more fundamentally) it looks like `bcl2fastq` might be using the expected barcode sequence from the manifest instead of the one that was actually read.  If that is true then using the value in the BC:Z tag from the read would not always give the right answer, and instead the BC: value stored in the header for the read group should be used instead (if available).

jenniferliddle commented:
> I'm not sure if this works properly yet. `samtools fastq -i --index-format 'i4i4' ...` doesn't write out the extra tags, but `samtools fastq -i -i1 index1.fq --index-format 'i4i4' ...` does.

So it doesn't. Now fixed.

> Also `samtools fastq -i ...` without `--index-format` does not complain about the lack of that option while it will if `-i1` is used.

Added a check to complain about this.

> When I do get them, the tags look like `1:N:0:ATCA-CGAT` on both reads, but [the bcl2fastq user manual](https://support.illumina.com/content/dam/illumina-support/documents/documentation/software_documentation/bcl2fastq/bcl2fastq2-v2-20-software-guide-15051736-03.pdf) says it should start with `2:` on read 2.

So it does! It was always supposed to use 1: or 2: This would appear to be a bug which went in with a PR from June of last year, but nobody noticed it. Now fixed.

> Also (and somewhat more fundamentally) it looks like `bcl2fastq` might be using the expected barcode sequence from the manifest instead of the one that was actually read. If that is true then using the value in the BC:Z tag from the read would not always give the right answer, and instead the BC: value stored in the header for the read group should be used instead (if available).

The BC:Z tag from the read is the *actual* barcode found, rather than the expected one, and therefore, I would argue, is the "right" one. Also, as far as I can see, we are not putting a BC value in the read group line of the header.
